crates/proto/src/xfer/retry_dns_handle.rs

```diff
@@ -94,7 +96,14 @@ where
             match self.stream.poll_next_unpin(cx) {
                 Poll::Ready(Some(Err(e))) => {
                     if self.remaining_attempts == 0 || !e.should_retry() {
-                        return Poll::Ready(Some(Err(e)));
+                        // Return the error originally returned from the highest priority name
+                        // server; if this is the first nameserver we've queried, return the error
+                        // we've just received.
+                        let error_to_return = self.first_error.clone().unwrap_or(e);
+                        return Poll::Ready(Some(Err(error_to_return)));
+                    }
+                    if let None = self.first_error {
+                        self.first_error = Some(e.clone());
```

Comment:
If you do this after decreasing `self.remaining_attempts` you can do it without cloning.

Reply:
good point, thanks!